frontend/src/app/modules/common/principal/principal-renderer.service.ts

```diff
@@ -0,0 +1,131 @@
+import {Injectable} from "@angular/core";
+import {PathHelperService} from "core-app/modules/common/path-helper/path-helper.service";
+import {ColorsService} from "core-app/modules/common/colors/colors.service";
+import {APIV3Service} from "core-app/modules/apiv3/api-v3.service";
+import {PrincipalHelper} from "core-app/modules/common/principal/principal-helper";
+import PrincipalType = PrincipalHelper.PrincipalType;
+
+export interface PrincipalLike {
+  id:string;
+  name:string;
+  href:string;
+}
+export interface AvatarOptions {
+  classes:string;
+}
+
+@Injectable({ providedIn: 'root' })
+export class PrincipalRendererService {
+
+  constructor(private pathHelper:PathHelperService,
+              private apiV3Service:APIV3Service,
+              private colors:ColorsService) {
+
+  }
+
+  renderMultiple(container:HTMLElement,
+                 users:PrincipalLike[],
+                 renderName:boolean = true,
+                 multiLine:boolean = false) {
+
+    const span = document.createElement('span');
+
+
+    for (let i = 0; i < users.length; i++) {
+      const avatar = document.createElement('span');
+      if (multiLine) {
+        avatar.classList.add('user-avatar--multi-line');
+      }
+
+      this.render(avatar, users[i], renderName);
+
+      if (!multiLine && i < users.length - 1) {
+        const sep = document.createElement('span');
+        sep.textContent = ', ';
+        avatar.appendChild(sep);
+      }
+
+      span.appendChild(avatar);
+    }
+
+    container.appendChild(span);
+  }
+
+  render(container:HTMLElement,
+         principal:PrincipalLike,
+         name:boolean = true,
+         avatar:false|AvatarOptions = { classes: 'avatar-medium' }):void {
+
+    const type = PrincipalHelper.typeFromHref(principal.href)!;
+
+    if (avatar) {
+      const el = this.renderAvatar(principal, avatar, type);
+      container.appendChild(el);
+    }
+
+    if (name) {
+      const el = this.renderName(principal, type);
+      container.appendChild(el);
+    }
+  }
+
+  private renderAvatar(principal:PrincipalLike, avatar:AvatarOptions, type:PrincipalType) {
+    const userInitials = this.getInitials(principal.name);
+    const colorCode = this.colors.toHsl(principal.name);
+
+    let fallback = document.createElement('div');
```

Comment:
Do we need to do manual rendering? Can we not just do a `<div [attr.style]="computedValue">`? That way we keep all html related stuff in the component. In fact, I'd argue we don't need this service at all. If anything, perhaps it's just a collection of pure functions that can be tested independently of anything angular, and then do the rest in the component itself.

Reply:
Hi @b12f , we need manual rendering for the table, where we are rendering outside of Angular. Something akin to a service is needed due to the injections to `PathHelper` and the other service. Of course this could also be a simple class/helper that you pass the injector to, but I don't see the advantage.